### PR TITLE
[slack] Share from Editor section

### PR DESCRIPTION
**Pull request by Harshg999**
## What changes were proposed in this pull request?

- Added share from editor section in gist modal
- Added send message API
- Update API UTs
- Link unfurling enabled for bot 

## How was this patch tested?

- Successfully running UTs and manually testing

Gist modal with changes and polished UI

Screenshot 2021-04-28 at 4 56 33 PM

On pressing `Post Message`

Screenshot 2021-04-28 at 4 59 47 PM

### CODE REVIEW

File: desktop/core/src/desktop/js/api/apiHelper.js [hunk 1]
```diff
@@ -1466,6 +1466,44 @@ class ApiHelper {
     });
   }
 
+  /**
+   *
+   * @param {Object} options
+   * @param {channel} options.channel
+   * @param {message} options.message
+   *
+   * @return {CancellableJqPromise<string>}
```
Comment (JohanAhlen) [inline desktop/core/src/desktop/js/api/apiHelper.js:1475]: `@return {Promise<void>}`

File: desktop/core/src/desktop/js/api/apiHelper.js [hunk 2]
```diff
@@ -1466,6 +1466,44 @@ class ApiHelper {
...
+   */
+  async sendSlackMessageAsync(options) {
+    const data = {
+      channel: options.channel,
+      message: options.message
+    };
+    return new Promise((resolve, reject) => {
+      simplePost(URLS.SEND_SLACK_MESSAGE, data, options)
+        .done(response => {
```
Comment (JohanAhlen) [inline desktop/core/src/desktop/js/api/apiHelper.js:1484]: nit: done(resolve)

File: desktop/core/src/desktop/js/api/apiHelper.js [hunk 3]
```diff
@@ -1466,6 +1466,44 @@ class ApiHelper {
...
+        })
+        .fail(reject);
+    });
+  }
+
+  /**
+   *
+   * @param {Object} options
+   *
+   * @return {CancellableJqPromise<string>}
```
Comment (JohanAhlen) [inline desktop/core/src/desktop/js/api/apiHelper.js:1495]: `Return {Promise<Object>}` ?

File: desktop/core/src/desktop/js/ko/components/ko.shareGistModal.js [hunk 4]
```diff
@@ -41,12 +43,52 @@ const TEMPLATE = `
...
+
+    this.link = ko.observable(params.link);
+    this.channels = ko.observableArray(params.channels);
+
+    this.selectedChannel = ko.observable('');
+    this.messageDescription = ko.observable('');
+  }
+
+  async postMessage() {
+    await ApiHelper.sendSlackMessageAsync({
```
Comment (JohanAhlen) [inline desktop/core/src/desktop/js/ko/components/ko.shareGistModal.js:85]: Surround with try catch to handle errors

File: desktop/core/src/desktop/js/ko/components/ko.shareGistModal.js [hunk 5]
```diff
@@ -41,12 +43,52 @@ const TEMPLATE = `
         </button>
```
Comment (JohanAhlen) [inline desktop/core/src/desktop/js/ko/components/ko.shareGistModal.js:43]: The panel will look nicer if the copy button is on the right of the text input for the gist link.

Comment (Harshg999) [reply desktop/core/src/desktop/js/ko/components/ko.shareGistModal.js:43]: Something like this @JohanAhlen ? or down right (currently its down left for reference) and not inline?

Screenshot 2021-04-28 at 1 02 01 PM

Comment (JohanAhlen) [reply desktop/core/src/desktop/js/ko/components/ko.shareGistModal.js:43]: Yes, I think that's better with the new form below. Thanks!

Comment (Harshg999) [reply desktop/core/src/desktop/js/ko/components/ko.shareGistModal.js]: Changes done!

**Review by JohanAhlen — COMMENTED** [review]
Nice! UI looks good to me. Will let @romainr comment on the backend changes.

File: desktop/core/src/desktop/lib/botserver/api.py [hunk 6]
```diff
@@ -43,3 +44,20 @@ def get_channels(request):
   return JsonResponse({
     'channels': bot_channels,
   })
+
+@api_error_handler
+def send_message(request):
+  channel = request.POST.get('channel')
+  message = request.POST.get('message')
+
+  slack_response = _send_message(channel, message)
```
Comment (romainr) [inline desktop/core/src/desktop/lib/botserver/api.py:53]: How about adding the name of the user? (as it seems that the author is shown-up is the SQL Bot, which is probably confusing)

Comment (romainr) [reply desktop/core/src/desktop/lib/botserver/api.py:53]: (especially that we don't check if a user has perms to post to the Slack channel, is there a way/API to validate that?)

Also, when listing channels, I guess we list all the public ones as the hue bot user, but that's fine for now.

File: desktop/core/src/desktop/js/ko/components/ko.shareGistModal.js [hunk 7]
```diff
@@ -32,21 +34,64 @@ const TEMPLATE = `
...
+          <input id="gistLink" style="width: 510px" ${ window.PREVENT_AUTOFILL_INPUT_ATTRS } onfocus="this.select()" data-bind="value: link" type="text" placeholder="${ I18n('Link') }"/>
+          <button class="btn" type="button" data-dismiss="modal" data-clipboard-target="#gistLink" data-bind="clipboard"><i class="fa fa-clipboard"></i></button>
         </form>
-        <button class="btn" type="button" data-dismiss="modal" data-clipboard-target="#gistLink" data-bind="clipboard">
-          <i class="fa fa-clipboard"></i> ${ I18n('Copy') }
-        </button>
       </div>
     </div>
+    <!-- ko if: window.SHARE_TO_SLACK -->
+    <label class="checkbox"><input type="checkbox" data-bind="checked: showSlackSection"> ${ I18n('Directly share to Slack') } </input></label>
```
Comment (romainr) [inline desktop/core/src/desktop/js/ko/components/ko.shareGistModal.js:45]: `Share to Slack`?

File: desktop/core/src/desktop/js/ko/components/ko.shareGistModal.js [hunk 8]
```diff
@@ -32,21 +34,64 @@ const TEMPLATE = `
...
+    <!-- ko if: showSlackSection -->
+    <form class="form-horizontal">
+      <div class="control-group">
+        <label class="control-label">${I18n('Message body')}</label>
```
Comment (romainr) [inline desktop/core/src/desktop/js/ko/components/ko.shareGistModal.js:49]: `Message`?

File: desktop/core/src/desktop/js/ko/components/ko.shareGistModal.js [hunk 9]
```diff
@@ -32,21 +34,64 @@ const TEMPLATE = `
...
+        </div>
+      </div>
+      <div class="control-group">
+        <label class="control-label">${I18n('Slack Channel')}</label>
+        <div class="controls">
+          <select class="input-xlarge" data-bind="options: channels, optionsCaption: 'Choose...', value: selectedChannel"></select>
+        </div>
+      </div>
+    </form>
+    <button class="btn" type="button" data-dismiss="modal" data-bind="click: postMessage">${I18n('Post Message')}</button>
```
Comment (romainr) [inline desktop/core/src/desktop/js/ko/components/ko.shareGistModal.js:61]: `Send`

Also possible to align button with message and channel?

File: desktop/core/src/desktop/js/ko/components/ko.shareGistModal.js [hunk 10]
```diff
@@ -32,21 +34,64 @@ const TEMPLATE = `
...
+    <!-- ko if: showSlackSection -->
+    <form class="form-horizontal">
+      <div class="control-group">
+        <label class="control-label">${I18n('Message body')}</label>
+        <div class="controls">
+          <input type="text" class="input-xlarge" data-bind="value: messageDescription, valueUpdate:'afterkeydown'" placeholder="${ I18n('(Optional)') }"/>
+        </div>
+      </div>
+      <div class="control-group">
+        <label class="control-label">${I18n('Slack Channel')}</label>
```
Comment (romainr) [inline desktop/core/src/desktop/js/ko/components/ko.shareGistModal.js:55]: `Channel`?

**Review by romainr — APPROVED** [review]
Nice, pretty clean and looks good!

File: desktop/core/src/desktop/js/ko/components/ko.shareGistModal.js (hunk 9 above)
Comment (Harshg999) [reply desktop/core/src/desktop/js/ko/components/ko.shareGistModal.js:61]: Made it inline-right with dropdown channel and message input length is same now @romainr 

Screenshot 2021-04-30 at 4 35 48 PM

File: desktop/core/src/desktop/lib/botserver/api.py (hunk 6 above)
Comment (romainr) [reply desktop/core/src/desktop/lib/botserver/api.py:53]: :+1:

**Pull request merged**